torch_xla/csrc/runtime/distributed_runtime.h

```diff
@@ -3,33 +3,78 @@
 
 #include <memory>
 
+#include "torch_xla/csrc/runtime/debug_macros.h"
+#include "tsl/distributed_runtime/preemption/preemption_sync_manager.h"
 #include "xla/pjrt/distributed/distributed.h"
 
 namespace torch_xla {
 namespace runtime {
 
+// DistributedRuntime serves as the point of entry for all operations which
+// required the XLA distributed runtime, such as preemption coordination.
 class DistributedRuntime {
  public:
-  static const std::string default_coordinator_port;
-  static DistributedRuntime& getInstance(int global_rank,
-                                         std::string master_addr,
-                                         std::string port) {
-    static DistributedRuntime dist_runtime_instance(global_rank, master_addr,
-                                                    port);
-    return dist_runtime_instance;
+  static inline const std::string kDefaultCoordinatorPort = "8547";
+
+  // Returns true if the distributed runtime has already been initialized.
+  static bool IsInitialized() { return dist_runtime_ != nullptr; }
+
+  // Initialize the shared DistributedRuntime object. This creates a
+  // DistributedRuntimeClient on each worker, and on global_rank 0 initializes
+  // the corresponding DistributedRuntimeService.
+  static void Initialize(int global_rank, int world_size,
+                         std::string master_addr, std::string port) {
+    XLA_CHECK(!IsInitialized()) << "DistributedRuntime already initialized";
+    dist_runtime_ = std::unique_ptr<DistributedRuntime>(
+        new DistributedRuntime(global_rank, world_size, master_addr, port));
+  }
+
+  // Shutdown the distributed runtime. All associated resources will be
+  // released, and subsequent calls to IsInitialized will return false.
+  // The distributed runtime may later be re-initialized.
+  static void Shutdown() {
+    XLA_CHECK(IsInitialized())
+        << "Must initialize distributed runtime before shutdown";
+    dist_runtime_ = nullptr;
+  }
+
+  // Retrieve the shared DistributedRuntime object.
+  static DistributedRuntime& Get() {
+    XLA_CHECK(IsInitialized())
+        << "Must initialize distributed runtime before retrieval";
+    return *dist_runtime_;
   }
```

Comment:
I wonder why we make `std::unique_ptr<DistributedRuntime>` a unique_ptr instead of just a DistributedRuntime&

Reply:
The `unique_ptr` makes ownership explicit, and it will manage destructing the singleton instance when it releases ownership (through reassignment or when it is destructed).